### PR TITLE
Fixed this socket is closed

### LINKED ISSUE #1473: This socket is closed
Related to #1415.

We've found the following steps:

1. Peer become unavailable.
2. Our server was closed by using `closeAll` method.
3. Our server socket was closed and it said `I was closed, ok.`
4. Server had a not empty list of sockets. These sockets are our `incoming` sockets. Server destroyed all sockets.
5. We didn't received any event on any `incoming` socket, because `incomingStream` is not created and `incoming` is not subscribed for any event.
6. Our pipe was created with dead `incoming` socket.
7. We are receiving error on `incoming` and only than we are destroying `incomingStream`.

So we need to do:

1. Subscribe to `incoming` immediately in order to not loose events.
2. Create `incomingStream` only if `incoming` is not closed.

### PR DESCRIPTION
**Pull request by andrew-aladev**
Closes #1473.

---
This change is [Reviewable](https://reviewable.io/reviews/thaliproject/thali_cordovaplugin/1474)

### CODE REVIEW

**Comment by ThaliTester** [comment]
PR is added to the queue for testing as 2. task. (22cf76c)

**Comment by andrew-aladev** [comment]
I will provide a simple test for this issue.

**Comment by ThaliTester** [comment]
Test (Fail) 92005191 build is completed (22cf76c)

See https://github.com/ThaliTester/TestResults/tree/9200519122cf76c_Fixed_`this_socket_is_closed`_andrew-aladev/ for the logs

**Comment by ThaliTester** [comment]
PR is added to the queue for testing as 4. task. (55ceef7)

**Comment by andrew-aladev** [comment]
Test done.

**Comment by ThaliTester** [comment]
PR is added to the queue for testing as 5. task. (ed1d7c6)

**Comment by ThaliTester** [comment]
Test (Fail) 92005191 build is completed (55ceef7)

See https://github.com/ThaliTester/TestResults/tree/9200519155ceef7_Fixed_this_socket_is_closed_andrew-aladev/ for the logs

**Comment by ThaliTester** [comment]
Test (Fail) 92005191 build is completed (ed1d7c6)

See https://github.com/ThaliTester/TestResults/tree/92005191ed1d7c6_Fixed_this_socket_is_closed_andrew-aladev/ for the logs

**Comment by yaronyg** [comment]
Once you fix the issue that makes all of our tests fail and decide if we do have a possible race condition then :lgtm: 

---

Reviewed 1 of 1 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[test/www/jxcore/bv_tests/testCreatePeerListener.js, line 932 at r2](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/1474#-KVgccuQajFYC5SeiaXh:-KVgccuQajFYC5SeiaXi:bewbuuq) ([raw file](https://github.com/thaliproject/thali_cordovaplugin/blob/ed1d7c66c34203cf0a1d417e057b18d0b1012ec9/test/www/jxcore/bv_tests/testCreatePeerListener.js#L932)):*
> ```JavaScript
>     var oldPipe = socket.__proto__.pipe;
>     socket.__proto__.pipe = function () {
>       t.fail('we should not create pipe with closed socket');
> ```

Please restore oldPipe. This code has replaced socket.__proto__.pipe for everyone so the next tests fail because they get a t.fail()!

---

*[test/www/jxcore/bv_tests/testCreatePeerListener.js, line 951 at r2](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/1474#-KVg_k-uHpEhE6X5z9vX:-KVg_k-vJYFZYrUrL6vE:b2s5jp1) ([raw file](https://github.com/thaliproject/thali_cordovaplugin/blob/ed1d7c66c34203cf0a1d417e057b18d0b1012ec9/test/www/jxcore/bv_tests/testCreatePeerListener.js#L951)):*
> ```JavaScript
>   });
> 
>   setUp(t, serversManager, applicationPort, nativePort, false, false,
> ```

Short Version: I think setUp should be called in the callback to nativeServer.listen.

Long Version: Isn't there a possible race condition where we call setUp who calls startAdvertisingAndListening which emits a peerAvailabilityChanged event which is handled by waitForPeerAvailabilityChanged who calls createPeerListener who then calls the code below that calls net.createConnection which will trigger a connect call to nativePort which might not be ready yet?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/1474#-:-KVgckKMJnN9hbQhRxjv:b-su73if)*

**Comment by ThaliTester** [comment]
PR is added to the queue for testing as 1. task. (e79faae)

**Comment by andrew-aladev** [comment]
Review status: 1 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[test/www/jxcore/bv_tests/testCreatePeerListener.js, line 932 at r2](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/1474#-KVgccuQajFYC5SeiaXh:-KVilore6w0SVZrBTGeE:b-896fix) ([raw file](https://github.com/thaliproject/thali_cordovaplugin/blob/ed1d7c66c34203cf0a1d417e057b18d0b1012ec9/test/www/jxcore/bv_tests/testCreatePeerListener.js#L932)):*
<details><summary><i>Previously, yaronyg (Yaron Y Goland) wrote…</i></summary>
> Please restore oldPipe. This code has replaced socket.__proto__.pipe for everyone so the next tests fail because they get a t.fail()!
</details>
Done.

---

*[test/www/jxcore/bv_tests/testCreatePeerListener.js, line 951 at r2](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/1474#-KVg_k-uHpEhE6X5z9vX:-KVilqGPWevLYV5B9suI:b-896fix) ([raw file](https://github.com/thaliproject/thali_cordovaplugin/blob/ed1d7c66c34203cf0a1d417e057b18d0b1012ec9/test/www/jxcore/bv_tests/testCreatePeerListener.js#L951)):*
<details><summary><i>Previously, yaronyg (Yaron Y Goland) wrote…</i></summary>
> Short Version: I think setUp should be called in the callback to nativeServer.listen.
> 
> Long Version: Isn't there a possible race condition where we call setUp who calls startAdvertisingAndListening which emits a peerAvailabilityChanged event which is handled by waitForPeerAvailabilityChanged who calls createPeerListener who then calls the code below that calls net.createConnection which will trigger a connect call to nativePort which might not be ready yet?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/1474)*

**Comment by ThaliTester** [comment]
Test (Success) 92005191 build is completed (e79faae)

See https://github.com/ThaliTester/TestResults/tree/92005191e79faae_Fixed_this_socket_is_closed_andrew-aladev/ for the logs

**Comment by ThaliTester** [comment]
Test 92005191e79faae(e79faae) has failed

See https://github.com/ThaliTester/TestResults/tree/92005191e79faae_Fixed_this_socket_is_closed_andrew-aladev/ for the fail logs
